src/mlpack/methods/ann/layer/residual_impl.hpp

```diff
@@ -0,0 +1,209 @@
+
+  // Note that we use `output` for the last layer; layerOutputs is only used for
+  // intermediate values between layers.
+  if (this->network.size() > 1)
+  {
+    // Initialize memory for the forward pass (if needed).
+    this->InitializeForwardPassMemory(input.n_cols);
+
+    for (size_t i = 0; i < this->network.size(); i++)
+      this->network[i]->Forward(input, this->layerOutputs[i]);
+
+		// Reduce the outputs to single output.
+    output.zeros();
+		for (size_t i = 0; i < this->layerOutputs.size(); i++)
+		{
+			output += this->layerOutputs[i];
+		}
+  }
+  else if (this->network.size() == 1)
+  {
+    this->network[0]->Forward(input, output);
+  }
+  else
+  {
+    // Empty network?
+    output = input;
+  }
+}
+
+template<typename MatType>
+void ResidualType<MatType>::Backward(
+    const MatType& input, const MatType& gy, MatType& g)
+{
+  if (this->network.size() > 1)
+  {
+    // Initialize memory for the backward pass (if needed).
+    this->InitializeBackwardPassMemory(input.n_cols);
+
+    g.zeros();
+    for (size_t i = 0; i < this->network.size(); i++) {
+      this->network[i]->Backward(this->layerOutputs[i], gy, this->layerDeltas[i]);
+			g += this->layerDeltas[i];
+		}
+  }
+  else if (this->network.size() == 1)
+  {
+    this->network[0]->Backward(input, gy, g);
+  }
+  else
+  {
+    // Empty network?
+    g = gy;
+  }
+}
+
+template<typename MatType>
+void ResidualType<MatType>::Gradient(
+    const MatType& input, const MatType& error, MatType& gradient)
+{
+  // We assume gradient has the right size already.
+
+  // Pass gradients through each layer.
+  if (this->network.size() > 1)
+  {
+    // Initialize memory for the gradient pass (if needed).
+    this->InitializeGradientPassMemory(gradient);
+
+    for (size_t i = 0; i < this->network.size(); ++i)
+    {
+      this->network[i]->Gradient(input, error, this->layerGradients[i]);
+    }
+  }
+  else if (this->network.size() == 1)
+  {
+    this->network[0]->Gradient(input, error, gradient);
+  }
+  else
+  {
+    // Nothing to do if the network is empty... there is no gradient.
+  }
+}
+
+template<typename MatType>
+void ResidualType<MatType>::ComputeOutputDimensions()
+{
+  this->inSize = 0;
+  this->totalInputSize = 0;
+  this->totalOutputSize = 0;
+
+  // Propagate the input dimensions forward to the output.
+  if (this->network.size() == 0)
+  {
+    this->outputDimensions = this->inputDimensions;
+    return;
+  }
+  this->inSize = this->inputDimensions[0];
+  for (size_t i = 1; i < this->inputDimensions.size(); ++i)
+    this->inSize *= this->inputDimensions[i];
+  this->totalInputSize = this->network.size() * this->inSize;
+
+  for (size_t i = 0; i < this->network.size(); ++i)
+  {
+    this->network[i]->InputDimensions() = this->inputDimensions;
+    size_t layerOutputSize = this->network[i]->OutputSize();
+    this->totalOutputSize += layerOutputSize;
+  }
+
+	// Compute the output size of the network using reduction rules.
```

Comment:
I can't tell in the Github diff view here, but is this a tab character here?  It would be worth going through the PR and making sure it uses two-space indentation with spaces, not tabs, to match the style guide. :+1: